File: apps/agent/src/agents/workspaceAgent/tools/listTasks.ts
import { FunctionTool } from '@google/adk';
import type { TaskProjection } from '@lifecoach/shared-types';
import { z } from 'zod';
import type { WorkspaceTokensStore } from '../../../storage/workspaceTokens.js';
import type { ExecFileLike } from '../gwsExec.js';
import { projectTask } from '../projections/task.js';
import { type RunGwsLogEvent, runGws } from '../runGws.js';

/**
 * `list_tasks` — Google Tasks list, filtered to needsAction by default.
 */

export const LIST_TASKS_TOOL_NAME = 'list_tasks';

export interface CreateListTasksToolDeps {
  store: WorkspaceTokensStore;
  uid: string;
  execFile?: ExecFileLike;
  log?: (event: RunGwsLogEvent) => void;
}

export type ListTasksResult =
  | { status: 'ok'; tasks: TaskProjection[]; truncated?: boolean }
  | { status: 'error'; code: string; message: string };

const parameters = z.object({
  taskListId: z.string().optional().describe('Task list id. Default "@default".'),
  showCompleted: z.boolean().optional().describe('Include completed tasks. Default false.'),
});

interface TasksListResponse {
  items?: unknown[];
}

export function createListTasksTool(deps: CreateListTasksToolDeps): FunctionTool {
  const { store, uid, execFile, log } = deps;

  // biome-ignore lint/suspicious/noExplicitAny: zod nominal mismatch with ADK generics
  return new FunctionTool<any>({
    name: LIST_TASKS_TOOL_NAME,
    description:
      'List Google Tasks in a task list. Returns projected task shapes (title, due, status, notes). Read-only.',
    parameters,
    execute: async (input: unknown): Promise<ListTasksResult> => {
      const args = input as { taskListId?: string; showCompleted?: boolean };
      const taskListId = args.taskListId ?? '@default';

      const result = await runGws({
        store,
        uid,
        toolName: LIST_TASKS_TOOL_NAME,
        service: 'tasks',
        resource: 'tasks',
        method: 'list',
        params: {
          tasklist: taskListId,
          showCompleted: args.showCompleted ?? false,
        },
        execFile,
        log,
      });
      if (result.status === 'error') {
        return { status: 'error', code: result.code, message: result.message };
      }
      const body = (result.body as TasksListResponse | null) ?? {};
      const tasks = (body.items ?? []).map((raw) =>
        // biome-ignore lint/suspicious/noExplicitAny: gws returns dynamic JSON
        projectTask(raw as any, taskListId),
      );
      return result.truncated ? { status: 'ok', tasks, truncated: true } : { status: 'ok', tasks };
    },
  });
}

File: apps/agent_py/src/lifecoach_agent/workspace_agent/tools/list_tasks.py
"""`list_tasks` — Google Tasks list, filtered to needsAction by default."""

from __future__ import annotations

from typing import Any

from lifecoach_agent.workspace_agent.projections import project_task
from lifecoach_agent.workspace_agent.run_gws import RunGwsOk, run_gws
from lifecoach_agent.workspace_agent.tools._deps import WorkspaceToolDeps

LIST_TASKS_TOOL_NAME = "list_tasks"


def create_list_tasks_tool(deps: WorkspaceToolDeps) -> Any:
    async def list_tasks(
        taskListId: str = "@default", showCompleted: bool = False
    ) -> dict[str, Any]:  # noqa: N803
        """List Google Tasks in a task list. Returns projected task
        shapes (title, due, status, notes). Read-only.

        Args:
            taskListId: Task list id. Default "@default".
            showCompleted: Include completed tasks. Default false.
        """
        result = await run_gws(
            store=deps.store,
            uid=deps.uid,
            tool_name=LIST_TASKS_TOOL_NAME,
            service="tasks",
            resource="tasks",
            method="list",
            params={"tasklist": taskListId, "showCompleted": showCompleted},
            build_client=deps.build_client,
            log=deps.log,
        )
        if not isinstance(result, RunGwsOk):
            return {"status": "error", "code": result.code, "message": result.message}
        body = result.body if isinstance(result.body, dict) else {}
        tasks = [
            project_task(item, taskListId).model_dump(by_alias=True, exclude_none=True)
            for item in (body.get("items") or [])
            if isinstance(item, dict)
        ]
        out: dict[str, Any] = {"status": "ok", "tasks": tasks}
        if result.truncated:
            out["truncated"] = True
        return out

    from google.adk.tools.function_tool import FunctionTool

    list_tasks.__name__ = LIST_TASKS_TOOL_NAME
    return FunctionTool(list_tasks)
